benchmark/common.js

```diff
@@ -88,52 +88,157 @@ Benchmark.prototype._queue = function(options) {
-      'benchmarks. Check benchmark/README.md for further instructions.');
-    process.exit(1);
-  }
+function AutocannonBenchmarker() {
+  const autocannon_exe = process.platform === 'win32'
+                         ? 'autocannon.cmd'
+                         : 'autocannon';
+  this.present = function() {
+    var result = child_process.spawnSync(autocannon_exe, ['-h']);
+    if (result.error && result.error.code === 'ENOENT')
+      return false;
+    else
+      return true;
+  };
+  this.create = function(path, duration, connections) {
+    const args = ['-d', duration, '-c', connections, '-j', '-n',
+                  'http://127.0.0.1:' + exports.PORT + path ];
+    var child = child_process.spawn(autocannon_exe, args);
+    child.stdout.setEncoding('utf8');
+    return child;
+  };
+  this.processResults = function(output) {
+    let result;
+    try {
+      result = JSON.parse(output);
+    } catch (err) {
+      // Do nothing, let next line handle this
+    }
+    if (!result || !result.requests || !result.requests.average) {
+      return undefined;
+    } else {
+      return result.requests.average;
+    }
+  };
 }
 
-// benchmark an http server.
-const WRK_REGEXP = /Requests\/sec:[ \t]+([0-9\.]+)/;
-Benchmark.prototype.http = function(urlPath, args, cb) {
-  hasWrk();
-  const self = this;
-
-  const urlFull = 'http://127.0.0.1:' + exports.PORT + urlPath;
-  args = args.concat(urlFull);
-
-  const childStart = process.hrtime();
-  const child = child_process.spawn('wrk', args);
-  child.stderr.pipe(process.stderr);
+function WrkBenchmarker() {
+  this.present = function() {
+    var result = child_process.spawnSync('wrk', ['-h']);
+    if (result.error && result.error.code === 'ENOENT')
+      return false;
+    else
+      return true;
+  };
+  this.create = function(path, duration, connections) {
+    const args = ['-d', duration, '-c', connections, '-t', 8,
+                  'http://127.0.0.1:' + exports.PORT + path ];
+    var child = child_process.spawn('wrk', args);
+    child.stdout.setEncoding('utf8');
+    child.stderr.pipe(process.stderr);
+    return child;
+  };
+  const regexp = /Requests\/sec:[ \t]+([0-9\.]+)/;
+  this.processResults = function(output) {
+    const match = output.match(regexp);
+    const result = match && +match[1];
+    if (!result)
+      return undefined;
+    else
+      return result;
+  };
+}
 
-  // Collect stdout
-  let stdout = '';
-  child.stdout.on('data', (chunk) => stdout += chunk.toString());
+const HTTPBenchmarkers = {
+  autocannon: new AutocannonBenchmarker(),
+  wrk: new WrkBenchmarker()
+};
 
-  child.once('close', function(code) {
-    const elapsed = process.hrtime(childStart);
-    if (cb) cb(code);
+// Benchmark an http server.
+Benchmark.prototype.http = function(urlPath, duration, connections, cb) {
+  const self = this;
+  duration = 1;
+
+  const picked_benchmarker = process.env.NODE_HTTP_BENCHMARKER ||
```

Comment:
@bzoz please see #8139, the issue is that `http/simple.js` takes too long time, using two benchmarks would make it take twice as long.

@mcollina 

> a) we want to be able to run the http benchmarks on Windows too, and it seem extremely hard to get wrk on Windows

I'm all for adding autocannon for getting Windows support.

> d) @mscdex and @jbergstroem argued that we are introducing a dependency on an installed version of Node, which might influence benchmarks (currently it is not #7180 (comment))

That is mostly an issue in when continuously monitoring performance. When just comparing a benchmark between master and a PR, that shouldn't be a problem.

> @AndreasMadsen would you mind reviewing if autocannon and wrk can be functionally equivalent?

It's actually impossible from a philosophy of science perspective to show this, however we can validate it.

First I run the `http/simple.js` using `wrk` and then using `autocannon`. (raw data https://gist.github.com/AndreasMadsen/619e4a447b8df7043f32771e64b7693f). To compare `wrk` with `autocannon` we need to compensate for the parameters, as there are few of them and we have many observations we can do that by using factors (the settings becomes a set of binary numbers) and then use a linear regression. Simultaneously we can also check how much the benchmarker affects the results by also making it a factor.

```
Call:
lm(formula = rate ~ ., data = dat)

Residuals:
    Min      1Q  Median      3Q     Max 
-3749.9 -1446.7   110.9  1376.7  5350.7 

Coefficients:
                Estimate Std. Error t value Pr(>|t|)    
(Intercept)     14003.64     121.29 115.456  < 2e-16 ***
c500               60.51      85.76   0.706    0.481    
chunks1          -823.67     105.04  -7.841 6.95e-15 ***
chunks4         -2715.50     105.04 -25.852  < 2e-16 ***
length1024      -1689.93     105.04 -16.088  < 2e-16 ***
length102400    -7251.31     105.04 -69.034  < 2e-16 ***
type bytes      -2110.44      85.76 -24.607  < 2e-16 ***
benchmarker wrk    46.03      85.76   0.537    0.592    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 1993 on 2152 degrees of freedom
Multiple R-squared:  0.7521,	Adjusted R-squared:  0.7513 
F-statistic: 932.5 on 7 and 2152 DF,  p-value: < 2.2e-16
Raw
```

From this result we can see that **the benchmarker does not have a statistical significant effect on the output**. We can see this because there are no stars right to `benchmarker wrk`. Where is `benchmarker autocannon` some tends to ask. What happens is that `autocannon` is set to the default and we are simply measuring the difference in performance caused by wrk. From the results we see that wrk is 46.03 ops/sec faster, but it is not significant.

I guess one could do a more detailed analysis of the interactions between the benchmarker and the individual parameters, but those results are tricky to interpret. I would say that since we can't observe a statistical significant effect we should assume that the benchmarker doesn't matter.